### PR TITLE
[SERVICE-479] integrate with non directory apps

### PR DESCRIPTION
**Pull request by spcfran**
Ready for review. 

Still not 100% test coverage

### CODE REVIEW

File: src/provider/index.ts
Comment (pjbroadbent) [inline src/provider/index.ts:152]: Another edge case - app exists in directory and/or is running, but hasn't (yet?) imported the client module.

Very edge-case-y though, not a concern for the time being.

File: src/provider/MetadataStore.ts
Comment (bryangaleopenfin) [inline src/provider/MetadataStore.ts:1]: Cool, makes sense!

File: src/provider/controller/SelectorHandler.ts
Comment (bryangaleopenfin) [inline src/provider/controller/SelectorHandler.ts:75]: Suggest move this class to using constructor injection, rather than adding more parameter injection. At the moment it's a bit of a mix, and I'd like it if we could move towards constructor, which I think is what `injectify` recommend

File: src/provider/index.ts
Comment (bryangaleopenfin) [inline src/provider/index.ts:98]: 👍

File: src/provider/controller/IntentHandler.ts
Comment (bryangaleopenfin) [inline src/provider/controller/IntentHandler.ts]: Are we worried about the case where an App has multiple windows, and the listeners are on a different window to the one returned by findWindowByAppName?

File: src/provider/controller/IntentHandler.ts
Comment (bryangaleopenfin) [inline src/provider/controller/IntentHandler.ts]: Resolve before merging

File: src/provider/controller/IntentHandler.ts
Comment (bryangaleopenfin) [inline src/provider/controller/IntentHandler.ts]: Feels odd to do this check here, but in the same method to trust that non-running apps will register listeners. I feel like we should leave this to be caught by `ensureReadyToReceiveIntent`. And what if raiseWithTarget is called during the 5 seconds we allow for the intent listener to be added?

File: src/provider/model/Model.ts
Comment (bryangaleopenfin) [inline src/provider/model/Model.ts]: Suggest don't worry about ticket references other than for 'TODO's

**Review by bryangaleopenfin — COMMENTED** [review]
Review of implementation changes - will look at tests separately

File: src/provider/model/Model.ts
Comment (bryangaleopenfin) [inline src/provider/model/Model.ts:113]: Suggest keep public methods above private methods

**Review by bryangaleopenfin — COMMENTED** [review]
Tests look good. Would maybe add a line of whitespace between tests for readability, and might be too much for this PR, but I wonder if there's any way to reduce code duplication (the tests where we have two listeners and try different combinations of unsubscribing them jump out for me and something we'd want to de-duplicate)

File: src/provider/controller/IntentHandler.ts
Comment (spcfran) [reply src/provider/controller/IntentHandler.ts]: Yep @pjbroadbent and I have discussed about the apps-with-multiple-windows thing quite a bit, we decided to leave it for a future story, once we know a bit more about how the resolver deals with individual windows, and whether the concept of "target" may include window information or just app

File: src/provider/controller/IntentHandler.ts
Comment (spcfran) [reply src/provider/controller/IntentHandler.ts]: @pjbroadbent thoughts?

File: src/provider/controller/IntentHandler.ts
Comment (spcfran) [reply src/provider/controller/IntentHandler.ts]: That's an excellent point. Will delegate that check to `fireIntent` and eventually throw the `Timeout waiting for intent listener to be added` if appropriate.

File: src/provider/controller/IntentHandler.ts
Comment (bryangaleopenfin) [reply src/provider/controller/IntentHandler.ts]: Cool cool. @pjbroadbent - just want to confirm this is as captured in Jira as it needs to be before I hit 'resolve conversation'

File: src/provider/controller/IntentHandler.ts
Comment (pjbroadbent) [reply src/provider/controller/IntentHandler.ts]: Let's discuss, come up with some wording that makes sense.

File: src/provider/controller/SelectorHandler.ts
Comment (spcfran) [reply src/provider/controller/SelectorHandler.ts:75]: ✔ Noted as a nice to have tidy-up task

File: src/provider/index.ts
Comment (spcfran) [reply src/provider/index.ts:152]: Mmhh not sure I follow. The app/window calling `addIntentListener` surely has imported the client?

Or do you mean the app/window that is the target of `raiseIntent`? In which case I think the "wait for intent handshake" in `fireIntent` would cover this?

**Review by tomrobertsOF — APPROVED** [review]
All fine by me

File: src/provider/controller/IntentHandler.ts
Comment (pjbroadbent) [reply src/provider/controller/IntentHandler.ts]: I've added a story to cover the potential upcoming changes that we discussed (SERVICE-505), but wrote it in a "TBD if we do this or not" sense.

May be worth for the time being changing the above check to fetch all windows from the model and search for a matching window here(?).

Comment (pjbroadbent) [reply src/provider/controller/IntentHandler.ts]: Scratch that. Was thinking the issue was specific to "intents with target", it's a bit more generic than that - and could do with a small story rather than building into this PR.

Have added SERVICE-506 to cover the change.

File: src/provider/index.ts
Comment (pjbroadbent) [reply src/provider/index.ts:152]: Sorry, my bad. Was assuming that `identity` could be an app that isn't using the FDC3 client - but since this method only runs if the service receives a message _from_ identity, then that window must have the client imported.

**Review by bryangaleopenfin — APPROVED** [review]

**Review by pjbroadbent — APPROVED** [review]

**Pull request merged**
